### ISSUE TITLE
network hotplug: add monitoring process for nets

### ISSUE THREAD

**Issue by egernst** [kata-containers/runtime#170]
Related to https://github.com/kata-containers/runtime/issues/113

This will track the implementation of the monitoring process.  virtcontainer changes to be tracked in #113

**Comment by sboeuf** [comment]
@WeiZhang555 @egernst @guangxuli @miaoyq @bergwolf @laijs I need your input on what language I should use for this ?
I have been talking with @amshinde about this and we kind of agreed that such a simple program would worth the shot in C since it would not bring too much overhead. If we choose Go, the footprint is gonna be bigger since we're going to import the netlink package, but it is gonna be very simple to implement too.
Keep in mind that the extra footprint is added per pod, since we need only one watcher per netns(which is tied to the pod).
I am waiting your inputs before I start implementing this.

**Comment by egernst** [comment]
I don't have a strong opinion on this.  Will it be the only C living in the project?  IIRC there was similar debate on shim.  What was the major takeaway on that decision?

**Comment by amshinde** [comment]
@egernst Yes, this would be the only C component. The main reason for the go implementation of shim was to do with the grpc library in C not being maintained.

**Comment by sboeuf** [comment]
Yes the gRPC library was the first reason and the ecosystem being all about Golang was the second reason. But that being said, as a future improvement I think the shim might move to a C component too. But the thing is that the shim is already functional and this is not a priority, but I don't want to make the same mistake with this component, I'd rather go with the right language on the first implementation for this one.

**Comment by miaoyq** [comment]
> But the thing is that the shim is already functional and this is not a priority, but I don't want to make the same mistake with this component, I'd rather go with the right language on the first implementation for this one.

@sboeuf  Agree with you.

**Comment by sboeuf** [comment]
@miaoyq so you mean you would prefer C language for this binary, right ?

**Comment by miaoyq** [comment]
@sboeuf Yeah

**Comment by guangxuli** [comment]
@sboeuf Both hands are in favor of using C language.

**Comment by miaoyq** [comment]
@egernst @sboeuf  Cloud  you add network hotplug related code in [govmm](https://github.com/intel/govmm)? 
So that we can usd it in `virtcontainers`.

**Comment by miaoyq** [comment]
@sboeuf  Glad to do it, but I am not familiar with `QMP`. I will have a try, and hope to get your guidance in the process. :-)

**Comment by sboeuf** [comment]
@miaoyq oh no worries, I'll help whenever you need ;)

**Comment by miaoyq** [comment]
@sboeuf  Cool, Thanks!

**Comment by sboeuf** [comment]
@miaoyq have you started working on this ? Any question so far or help needed so far ?

**Comment by miaoyq** [comment]
>@miaoyq have you started working on this ? Any question so far or help needed so far ?

@sboeuf Yeah.
I read the code about network and found the CNI plugin was called in virtcontainers only for getting `DNS`, see:
https://github.com/kata-containers/runtime/blob/f74f61e8d1817a2775d5f8a70e4e2ad443e70821/virtcontainers/cni.go#L151

this will cause eth0 to be added twice(container runtime call CNI plugin firstly),  but the CNI spec includes the following description:
```
A runtime must not call ADD twice (without a corresponding DEL) for the same (network name, container id). 
In other words, a given container ID must be added to a specific network exactly once.
```
see https://github.com/containernetworking/cni/blob/master/SPEC.md#general-considerations
Is there any special purpose? I'm not very clear about this.

This also makes  kata-runtime dependent on the CNI plugin.

**Comment by sboeuf** [comment]
@miaoyq Don't worry about this because in case of kata-runtime, the network type used is `CNM`. Take a look into `runtime/virtcontainers/cnm.go` instead.
The purpose of this `runtime/virtcontainers/cni.go` file is out of the OCI spec scope.

**Comment by miaoyq** [comment]
> The purpose of this runtime/virtcontainers/cni.go file is out of the OCI spec scope

@egernst Does `k8s + cri-o + kata-runtime` also use `CNM` ?

**Comment by miaoyq** [comment]
> The purpose of this runtime/virtcontainers/cni.go file is out of the OCI spec scope

@sboeuf  Does `k8s + cri-o + kata-runtime` also use `CNM` ?

**Comment by sboeuf** [comment]
>Does k8s + cri-o + kata-runtime also use CNM ?

In short, yes.
But let me try to explain what we mean by CNM. When you use `kata-runtime`, you're dealing with an OCI compliant runtime, which relies on the CNM hooks for the network. That's why it always uses this implementation `cnm.go`, but this does not mean that CNI is not used. You have to think about `cnm.go` from the virtcontainers perspective where the scope was not only to support an OCI runtime on top.
Of course, when you use `k8s` + `cri-o` + `kata-runtime`, `k8s` will use `CNI` and the `cnm.go` implementation is going to be a simple passthrough because nothing will be provided.

**Comment by miaoyq** [comment]
@sboeuf Got it, Thank you for your detailed explanation. :)

**Comment by sboeuf** [comment]
No problem! I know this part is really not obvious...
